client/src/main/kotlin/io/spine/examples/chatspn/desktop/Navigation.kt

```diff
@@ -26,48 +26,692 @@
+    val userEmailState: MutableState<String> = mutableStateOf("")
+    val errorState: MutableState<Boolean> = mutableStateOf(false)
+}
+
+/**
+ * Converts `Timestamp` to the `hh:mm` string.
+ */
+public fun Timestamp.toStringTime(): String {
+    val date = Date(this.seconds * 1000)
+    val format = SimpleDateFormat("hh:mm", Locale.getDefault())
+    return format.format(date)
 }
 
 /**
  * Pages in the application.
  */
 private enum class Page {
-    CHATS, LOGIN, REGISTRATION
+    CHAT, LOGIN, REGISTRATION, PROFILE
+}
+
+/**
+ * Data for the chat preview.
+ */
+public data class ChatData(
+    val id: ChatId,
+    val name: String,
+    val lastMessage: MessageData?,
+    val members: List<UserId>,
+    val type: Chat.ChatType
+)
+
+/**
+ * List of `ChatData`.
+ */
+public typealias ChatList = List<ChatData>
+
+/**
+ * Creates the `ChatData` list from the `ChatPreview` list.
+ *
+ * @param client desktop client to find user profiles
+ */
+private fun List<ChatPreview>.toChatDataList(client: DesktopClient): ChatList {
+    return this.stream().map { chatPreview ->
+        ChatData(
+            chatPreview.id,
+            chatPreview.name(client),
+            chatPreview.lastMessageData(client),
+            chatPreview.members(client),
+            chatPreview.type()
+        )
+    }.collect(Collectors.toList())
+}
+
+/**
+ * Retrieves the display name of the chat.
+ *
+ * @param client desktop client to find user profiles
+ */
+private fun ChatPreview.name(client: DesktopClient): String {
+    if (this.hasGroupChat()) {
+        return this.groupChat.name
+    }
+    val creator = this.personalChat.creator
+    val member = this.personalChat.member
+    if (client.authenticatedUser!!.id.equals(creator)) {
+        return client.findUser(member)!!.name
+    }
+    return client.findUser(creator)!!.name
+}
+
+/**
+ * Retrieves the last message from the `ChatPreview` and creates the `MessageData` from it.
+ */
+private fun ChatPreview.lastMessageData(client: DesktopClient): MessageData? {
+    val isMessageDefault = this.lastMessage.equals(MessagePreview.getDefaultInstance())
+    return if (isMessageDefault) {
+        null
+    } else {
+        this.lastMessage.toMessageData(client)
+    }
+}
+
+/**
+ * Retrieves chat type.
+ */
+private fun ChatPreview.type(): Chat.ChatType {
+    return if (this.hasGroupChat()) {
+        Chat.ChatType.CT_GROUP
+    } else {
+        Chat.ChatType.CT_PERSONAL
+    }
+}
+
+/**
+ * Retrieves members of the chat.
+ */
+private fun ChatPreview.members(client: DesktopClient): List<UserId> {
+    return client.readChatMembers(this.id)
+}
+
+/**
+ * Creates the `MessageData` from the `MessagePreview`.
+ *
+ * @param client desktop client to find user profiles
+ */
+private fun MessagePreview.toMessageData(client: DesktopClient): MessageData {
+    return MessageData(
+        this.id,
+        client.findUser(this.user)!!,
+        this.content,
+        this.whenPosted
+    )
 }
```

Comment:
Same as above. 717 lines is already too long. I wouldn't mind a file of this size, if it was much more homogeneous. But this one splits more or less logically into much more compact parts.

Please do.